### PR TITLE
Ui filepicker

### PR DESCRIPTION
**Pull request by yanchith**
This adds another type of string parameter to the UI - a file path. The UI launches a filepicker when a file path string input is used.

@ondrowan This does not launch the filepicker in a separate thread on purpose. Janper assures me blocking all rendering is a perfectly ok user experience.

Items to do:

- [x] Test window resizing (to see if `calc_item_width` always does the correct thing)
- [x] Test on linux
- [x] Test on Windows
- [x] Test on macOS

### CODE REVIEW

File: src/ui.rs [hunk 1]
```diff
@@ -748,3 +757,56 @@ fn push_disabled_style(ui: &imgui::Ui) -> (imgui::ColorStackToken, imgui::StyleS
...
     (color_token, style_token)
 }
+
+fn file_input(
+    ui: &imgui::Ui,
+    label: &imgui::ImStr,
+    file_ext_filter: Option<(&str, &str)>,
+    buffer: &mut imgui::ImString,
+) -> bool {
+    let edit_button_label = imgui::im_str!("Edit##{}");
```
Comment (janper) [inline src/ui.rs]: "Open"

**Review by janper — CHANGES_REQUESTED** [review]
Technically it works, but the path looks like this: image

or this, in case of a relative path 
image

but if you the Edit button for the second time, it looks like this 
image and of course this doesn't work.

Also it's not possible to edit the path manually.

**Comment by yanchith** [comment]
@janper

> Technically it works, but the path looks like this

Didn't test on windows yet. Looks like a windows string encoding thing. I'll look into it.

> or this, in case of a relative path

That looks correct to me. Is there a problem in how the relative path looks?

> Also it's not possible to edit the path manually.

That's on purpose. Do we want it editable?

> but if you the Edit button for the second time, it looks like this

How did you manage to do that? :joy: EDIT: Ok, I see. Rookie mistake on my part.

**Comment by janper** [comment]
> That looks correct to me. Is there a problem in how the relative path looks?

I'd expect it to start with `.\`. On the other hand, relative paths are not so common on windows, so I guess there is no standard.

**Comment by janper** [comment]
> That's on purpose. Do we want it editable?

hmmm... if copypaste works, then probably yes. if it doesn't, then don't bother

**Comment by janper** [comment]
> How did you manage to do that? 😂 EDIT: Ok, I see. Rookie mistake on my part.

I tested thoroughly. Clicked the button twice

**Comment by yanchith** [comment]
> > That looks correct to me. Is there a problem in how the relative path looks?
> 
> I'd expect it to start with `.\`. On the other hand, relative paths are not so common on windows, so I guess there is no standard.

Relative paths without `./` are valid AFAIK, but it's not a problem to add if you think it's important.

**Comment by yanchith** [comment]
> hmmm... if copypaste works, then probably yes. if it doesn't, then don't bother

It doesn't. We have a low prio task to fix it.

**Comment by janper** [comment]
> It doesn't. We have a low prio task to fix it.

Then let's not do it

**Comment by janper** [comment]
> Relative paths without ./ are valid AFAIK, but it's not a problem to add if you think it's important.

Nah, let's not

**Comment by janper** [comment]
image

**Comment by yanchith** [comment]
~Yeah, the width of the button is hardcoded and I changed the content...~ Nope, it isn't hardcoded, I just forgot to use the computed width 😄

**Comment by janper** [comment]
yup, it's just this now 
image

**Review by janper — APPROVED** [review]
Great, thanks!

**Pull request merged**